added instagram channel code

Horizon733 opened the pull request:
**Proposed changes**:
- added channel file with working and tested code with sample bot I had also attaching issues link below to check outputs.
Please let me know what all changes are to be done since this is my first PR.

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

Horizon733 commented:
I added changelog, need help with code quality check, how to run tests and integration tests

TyDunn commented:
Sweet! I added the other issue it addresses to the description and asked an engineer to give you some guidance. Please make sure to target Rasa Open Source 3.0 since this is an enhancement (check out the repo README for more information)

Horizon733 commented:
> Sweet! I added the other issue it addresses to the description and asked an engineer to give you some guidance. Please make sure to target Rasa Open Source 3.0 since this is an enhancement (check out the repo README for more information)

Sure will focus on 3.0 but how do I test this with 3.0? And I read the whole Readme in main branch can you please tell which section was related to this?

Horizon733 commented:
> Sweet! I added the other issue it addresses to the description and asked an engineer to give you some guidance. Please make sure to target Rasa Open Source 3.0 since this is an enhancement (check out the repo README for more information)

Any update?

TyDunn commented:
> Any update?
Yes. Sorry for the delay. We have been really busy the last couple weeks. It looks like an engineer will take a look at this this week

alwx commented on rasa/core/channels/instagram.py:
```diff
@@ -0,0 +1,334 @@
+logger = logging.getLogger(__name__)
+
+
+class Messenger:
```
This class requires a docstring (see `facebook.py` for an example)

alwx replied:
In addition, the class name is the same as the one in `facebook.py`

Horizon733 replied:
I am on it, changing it to InstaMessenger and adding docstring wherever required

alwx commented on rasa/core/channels/instagram.py:
```diff
@@ -0,0 +1,334 @@
+        self.client = MessengerClient(page_access_token)
+        self.last_message: Dict[Text, Any] = {}
+
+    def get_user_id(self, metadata) -> Text:
```
The docstring is missing.

alwx commented on rasa/core/channels/instagram.py:
```diff
@@ -0,0 +1,334 @@
+        """Check if the message is a message from the user"""
+        print(
+            f"function-name: {inspect.stack()[0][3]}\n" f"user_message: {str(message)}"
+        )
```
Have you accidentally left printing for debugging purposes?

Horizon733 replied:
yes, it was added for debug purpose, will remove it

alwx reviewed (changes requested):
Hey @Horizon733 and thanks for your contribution!

I did a review and have a couple of questions for you:
1) It looks very very similar to the code in `rasa/core/channels/facebook.py` — I guess it's because Instagram uses Messenger as well. Can we somehow use the same Messenger class that we already have for Instagram conversations or at least write some generic class for both Instagram and Facebook?
2) Will it work for non-Messenger conversations? The Messenger update for Instagram seems to only be available for those who have their Instagram account linked to their Facebook account (https://about.fb.com/news/2020/09/new-messaging-features-for-instagram/); the rest use standard Instagram messaging.
3) Have you considered adding tests for this channel?

alwx commented:
@Horizon733 

To be honest, I don't think it's worth adding this class in the current form — the `rasa/core/channels/instagram.py` is basically just a duplicate for `rasa/core/channels/facebook.py`, and probably the right way of doing it is to just update the documentation and indicate that the existing `Messenger` class needs to be used for Instagram since Instagram uses Messenger anyway.
I got your point the Messenger in Instagram is still in beta but I think this aspect should be covered by tests.

Would be happy to hear more opinions on that. @TyDunn, @wochinge?

TyDunn commented:
@alwx I did not realize that the APIs were so integrated across Instagram and Facebook. If is just a matter of updating the docs and making slight changes to the existing Messenger channel (e.g. maybe changing the name from `facebook.py` to `messenger.py`) to make it clear and easy for users to use it with Instagram, then I agree that we should go that route

Horizon733 commented:
> @alwx I did not realize that the APIs were so integrated across Instagram and Facebook. If is just a matter of updating the docs and making slight changes to the existing Messenger channel (e.g. maybe changing the name from `facebook.py` to `messenger.py`) to make it clear and easy for users to use it with Instagram, then I agree that we should go that route

Yes, we can definitely do that, and add some checks for unique messages of facebook and Instagram. Let me know if anything is there, I could help out in changing facebook.py to messenger.py to keep it generic

rgstephens commented:
I agree with @alwx that a custom channel is not required for Instagram. I just brought up the Facebook channel with Instagram.

The [docs](https://developers.facebook.com/docs/messenger-platform/instagram/get-started) show that you use the existing Facebook Messenger channel. This is done by connecting the Rasa channel to a Facebook page. Integration with Instagram requires it to be linked to a Facebook page.

I think this PR should be closed along with #9446 and #9594.

Horizon733 commented:
> I agree with @alwx that a custom channel is not required for Instagram. I just brought up the Facebook channel with Instagram.
> 
> The [docs](https://developers.facebook.com/docs/messenger-platform/instagram/get-started) show that you use the existing Facebook Messenger channel. This is done by connecting the Rasa channel to a Facebook page. Integration with Instagram requires it to be linked to a Facebook page.
> 
> I think this PR should be closed along with #9446 and #9594.

Yes, that's why I am creating on solo channel which supports both, where it also checks if instagram is used then it will convert buttons to quick replies. Send it I not supported on Instagram

TyDunn commented:
@rgstephens We want to update the docs and make the small adjustment that @Horizon733 to make it clear to users

Horizon733 commented:
Also, I need help with committing my changes and pushing it here, but it is not able to get complete, I did used pre-commit there it shows me, 1 file left unchanged and that file I also try to add and commit but black isn't letting it to get commit.
I would really appreciate if someone could help.
Please let me know how to add changes to docs, will add that too

TyDunn commented:
@Horizon733: @alwx is busy getting the first 3.0 release candidate out right now, but he will get back to you next week 👍

Horizon733 commented:
> @Horizon733: @alwx is busy getting the first 3.0 release candidate out right now, but he will get back to you next week 👍

no issues, till then I will update docs

alwx commented:
@Horizon733 hey!
1) First of all, could you please reset `.pre-commit-config.yaml`?
2) In order to fix formatting I would suggest you to do `git rebase` — the idea is to first base your changes on the recent changes from the repo (we've merged a lot of stuff in the last weeks!). After you do this, you just need to run `make formatter` to apply formatting to the code. It's also a good idea to run `make lint` to check if everything is correct with types/etc.

Regarding the docs:
There are two files that need to be updated: `docs/docs/messaging-and-voice-channels.mdx` and `docs/docs/connectors/facebook-messenger.mdx`. I think the last one needs to be renamed (to something like `facebook-and-instagram.mdx`) and modified, while the first one only requires some small adjustments.

Horizon733 commented:
> @Horizon733 hey!
> 
> 1. First of all, could you please reset `.pre-commit-config.yaml`?
> 2. In order to fix formatting I would suggest you to do `git rebase` — the idea is to first base your changes on the recent changes from the repo (we've merged a lot of stuff in the last weeks!). After you do this, you just need to run `make formatter` to apply formatting to the code. It's also a good idea to run `make lint` to check if everything is correct with types/etc.
> 
> Regarding the docs: There are two files that need to be updated: `docs/docs/messaging-and-voice-channels.mdx` and `docs/docs/connectors/facebook-messenger.mdx`. I think the last one needs to be renamed (to something like `facebook-and-instagram.mdx`) and modified, while the first one only requires some small adjustments.

Thanks for helping out, rebase worked and conflicts has been solved locally
the final step which is `make lint` failed and below was the error.
```
rasa\jupyter.py:11:5: F401 'rasa.core.agent.Agent' imported but unused
rasa\telemetry.py:38:5: F401 'rasa.core.channels.channel.InputChannel' imported but unused
rasa\telemetry.py:42:5: F401 'rasa.core.utils.AvailableEndpoints' imported but unused
rasa\telemetry.py:897:5: F811 redefinition of unused 'AvailableEndpoints' from line 42
rasa\core\run.py:19:1: F401 'rasa.core.channels.channel.InputChannel' imported but unused
rasa\core\actions\action.py:73:5: F401 'rasa.shared.core.events.IntentPrediction' imported but unused
rasa\core\brokers\file.py:10:5: F401 'rasa.utils.endpoints.EndpointConfig' imported but unused
rasa\core\brokers\pika.py:18:1: F401 'rasa.utils.endpoints.EndpointConfig' imported but unused
rasa\core\featurizers\single_state_featurizer.py:121:47: F821 undefined name 'ignore'
rasa\core\featurizers\single_state_featurizer.py:121:47: F821 undefined name 'item'
rasa\core\featurizers\single_state_featurizer.py:125:44: F821 undefined name 'ignore'
rasa\core\featurizers\single_state_featurizer.py:125:44: F821 undefined name 'item'
rasa\core\training\training.py:10:5: F401 'rasa.shared.core.trackers.DialogueStateTracker' imported but unused
rasa\core\training\training.py:11:5: F401 'rasa.shared.core.events.Event' imported but unused
rasa\core\training\__init__.py:5:5: F401 'rasa.shared.core.generator.TrackerWithCachedStates' imported but unused
rasa\core\training\__init__.py:7:5: F401 'rasa.shared.importers.importer.TrainingDataImporter' imported but unused
rasa\core\training\__init__.py:72:5: F811 redefinition of unused 'TrainingDataImporter' from line 7
rasa\engine\graph.py:387:61: F821 undefined name 'ignore'
rasa\engine\graph.py:387:61: F821 undefined name 'no'
rasa\engine\graph.py:387:61: F821 undefined name 'redef'
rasa\engine\recipes\default_recipe.py:143:44: F821 undefined name 'RegisteredComponent'
rasa\nlu\extractors\crf_entity_extractor.py:41:5: F401 'sklearn_crfsuite.CRF' imported but unused
rasa\nlu\extractors\mitie_entity_extractor.py:31:5: F401 'mitie' imported but unused
rasa\nlu\extractors\mitie_entity_extractor.py:128:9: F811 redefinition of unused 'mitie' from line 31
rasa\nlu\extractors\mitie_entity_extractor.py:164:9: F811 redefinition of unused 'mitie' from line 31
rasa\nlu\extractors\mitie_entity_extractor.py:272:9: F811 redefinition of unused 'mitie' from line 31
rasa\nlu\tokenizers\spacy_tokenizer.py:11:5: F401 'spacy.tokens.doc.Doc' imported but unused
rasa\nlu\utils\bilou_utils.py:24:5: F401 'rasa.nlu.tokenizers.tokenizer.Token' imported but unused
rasa\nlu\utils\mitie_utils.py:13:5: F401 'mitie' imported but unused
rasa\nlu\utils\mitie_utils.py:25:9: F811 redefinition of unused 'mitie' from line 13
rasa\nlu\utils\mitie_utils.py:86:9: F811 redefinition of unused 'mitie' from line 13
rasa\shared\core\trackers.py:70:5: F401 'rasa.shared.core.events.NLUPredictionData' imported but unused
rasa\shared\core\training_data\loading.py:12:1: F401 'rasa.shared.core.training_data.structures.StoryStep' imported but unused
rasa\shared\core\training_data\visualization.py:23:5: F401 'rasa.shared.nlu.training_data.message.Message' imported but unused
rasa\shared\core\training_data\story_writer\story_writer.py:8:5: F401 'rasa.shared.core.events.Event' imported but unused
rasa\shared\core\training_data\story_writer\story_writer.py:9:5: F401 'rasa.shared.core.training_data.structures.StoryStep' imported but unused
rasa\shared\core\training_data\story_writer\story_writer.py:63:9: F811 redefinition of unused 'StoryStep' from line 9
rasa\shared\nlu\training_data\loading.py:19:5: F401 'rasa.shared.nlu.training_data.formats.readerwriter.TrainingDataReader' imported but unused
rasa\shared\nlu\training_data\message.py:29:5: F401 'rasa.shared.nlu.training_data.features.Features' imported but unused
rasa\utils\io.py:27:5: F401 'prompt_toolkit.validation.Validator' imported but unused
rasa\utils\io.py:174:5: F811 redefinition of unused 'Validator' from line 27
rasa\utils\train_utils.py:39:5: F401 'rasa.nlu.extractors.extractor.EntityTagSpec' imported but unused
rasa\utils\train_utils.py:40:5: F401 'rasa.nlu.tokenizers.tokenizer.Token' imported but unused
rasa\utils\train_utils.py:41:5: F401 'tensorflow.keras.callbacks.Callback' imported but unused
Makefile:84: recipe for target 'lint' failed
make: *** [lint] Error 1
```

Horizon733 commented:
Also @alwx , when I try committing it gives me below error.
```
[INFO] Installing environment for https://github.com/thlorenz/doctoc.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Restored changes from C:\Users\Dishant\.cache\pre-commit\patch1636636840-14288.
An unexpected error has occurred: CalledProcessError: command: ('C:\\Users\\Dishant\\.cache\\pre-commit\\repol613k9bc\\node_env-default\\Scripts\\npm.CMD', 'install', '--dev', '--prod', '--ignore-prepublish', '--no-progress', '--no-save')
return code: 3
expected return code: 0
stdout: (none)
stderr:
    OpenSSL configuration error:
    543C0000:error:80000003:system library:BIO_new_file:No such process:c:\ws\deps\openssl\openssl\crypto\bio\bss_file.c:67:calling fopen(F:\PostgreSQL\psqlODBC\etc\openssl.cnf, rb)
    543C0000:error:10000080:BIO routines:BIO_new_file:no such file:c:\ws\deps\openssl\openssl\crypto\bio\bss_file.c:75:
    543C0000:error:07000072:configuration file routines:def_load:no such file:c:\ws\deps\openssl\openssl\crypto\conf\conf_def.c:179:
    OpenSSL configuration error:
    B43C0000:error:80000003:system library:BIO_new_file:No such process:c:\ws\deps\openssl\openssl\crypto\bio\bss_file.c:67:calling fopen(F:\PostgreSQL\psqlODBC\etc\openssl.cnf, rb)
    B43C0000:error:10000080:BIO routines:BIO_new_file:no such file:c:\ws\deps\openssl\openssl\crypto\bio\bss_file.c:75:
    B43C0000:error:07000072:configuration file routines:def_load:no such file:c:\ws\deps\openssl\openssl\crypto\conf\conf_def.c:179:
Check the log at C:\Users\Dishant\.cache\pre-commit\pre-commit.log
```

TyDunn commented:
@Horizon733 have you tried with a clean virtual env or downgrading / upgrading your virtual environment?

Horizon733 commented:
Hi, @alwx @TyDunn please run review the code I have pushed, let me know if there are any changes

rlinke commented on README.md:
```diff
@@ -29,8 +29,7 @@ updates to 3.0. Read more about [our contributor guidelines](#how-to-contribute)
 Rasa is an open source machine learning framework to automate text-and voice-based conversations. With Rasa, you can build contextual assistants on:
-- Facebook Messenger
-- Slack
+- Facebook Messengcd - Slack
```
```suggestion
- Facebook Messenger
- Slack
```

Horizon733 replied:
fixed!

rlinke reviewed (commented):
small typo in the readme

rlinke commented:
Sorry to interrupt/join the discussion. I don't completely understand why the channels should be merged. 
**point 1** 
From feature side the IG API is limited compared to FB messenger API. Wouldn't handling these differences be much easier in 2 separated implementations?
I noticed the IG webhook returns a [echo message](https://developers.facebook.com/docs/messenger-platform/instagram/features/webhook/#echo-webhook) to the endpoint, which will be discarded with a warning
`Received a message from instagram that we can not handle. Message: {
   "sender":{
      "id":"<someid>"
   },
   "recipient":{
      "id":"<someotherid>"
   },
   "timestamp":<msg_ts>,
   "message":{
      "mid":"<some_mid>",
      "text":"<original_bot_message>",
      "is_echo":true
   }
}`
With <original_bot_message> being the message I just sent from rasa to the IG user.

**point 2**
I tried to setup IG/Facebook permissions, however I end up with two separate/distinct PAT (page-access-tokens) for FB messenger and IG on a linked IG account to a linked page (i.e. my test Facebook page for facebook messenger has a different PAT than my test Instagram channel, both accounts seem to be properly linked). 
For this I would need to create to separate configs in my endpoints.yml or be able to add both PAT to the one channel.

I tested by following @Horizon733 suggestion by duplicating the facebook.py and make the IG changes (mostly by exchanging "facebook" --> "instagram"). Afterwards I added the seperate instagram channel config in the endpoints.yml.

With this I could successfully talk from facebook messenger and from instagram chat with the rasa chatbot.

Thoughts? @Horizon733  @alwx

rgstephens commented:
> I noticed the IG webhook returns a [echo message](https://developers.facebook.com/docs/messenger-platform/instagram/features/webhook/#echo-webhook) to the endpoint, which will be discarded with a warning `Received a message from instagram that we can not handle. Message: { "sender":{ "id":"<someid>" }, "recipient":{ "id":"<someotherid>" }, "timestamp":<msg_ts>, "message":{ "mid":"<some_mid>", "text":"<original_bot_message>", "is_echo":true } }` With <original_bot_message> being the message I just sent from rasa to the IG user.

I did not run into this warning but this issue could be handled without the new for a separate channel.

> **point 2** I tried to setup IG/Facebook permissions, however I end up with two separate/distinct PAT (page-access-tokens) for FB messenger and IG on a linked IG account to a linked page (i.e. my test Facebook page for facebook messenger has a different PAT than my test Instagram channel, both accounts seem to be properly linked). For this I would need to create two separate channel configs in my endpoints.yml or be able to add both PAT to the one channel.

I found that I needed to create the second PAT for IG but because IG is linked to the Facebook page my channel config still must use the FB PAT and Rasa messages are received by the Facebook page that is linked to IG.

rlinke commented:
> I did not run into this warning but this issue could be handled without the need for a separate channel.
 
ok yes I agree on this. Esp. I noticed, that the web browser IG chat has limited capabilities and I wrongly assumed that e.g. quick replies would work differently as they were not shown in the web version. On the app it works as expected with the default code.

> I found that I needed to create the second PAT for IG but because IG is linked to the Facebook page my channel config still must use the FB PAT and Rasa messages are received by via the Facebook page that is linked to IG.
ok I will test this out, to make sure, how to do it. 
you basically use "any recent" rasa with the vanilla facebook channel, but with your linked IG account --> the singular facebook webhook handles the IG messages?

Horizon733 commented:
> Sorry to interrupt/join the discussion. I don't completely understand why the channels should be merged. **point 1** From feature side the IG API is limited compared to FB messenger API. Wouldn't handling these differences be much easier in 2 separated implementations? E.g. I noticed the IG webhook returns a [echo message](https://developers.facebook.com/docs/messenger-platform/instagram/features/webhook/#echo-webhook) to the endpoint, which will be discarded with a warning `Received a message from instagram that we can not handle. Message: { "sender":{ "id":"<someid>" }, "recipient":{ "id":"<someotherid>" }, "timestamp":<msg_ts>, "message":{ "mid":"<some_mid>", "text":"<original_bot_message>", "is_echo":true } }` With <original_bot_message> being the message I just sent from rasa to the IG user.
> 
> **point 2** I tried to setup IG/Facebook permissions, however I end up with two separate/distinct PAT (page-access-tokens) for FB messenger and IG on a linked IG account to a linked page (i.e. my test Facebook page for facebook messenger has a different PAT than my test Instagram channel, both accounts seem to be properly linked). For this I would need to create two separate channel configs in my endpoints.yml or be able to add both PAT to the one channel.
> 
> I tested by following @Horizon733 suggestion by duplicating the facebook.py and make the IG changes (mostly by exchanging "facebook" --> "instagram"). Afterwards I added the seperate instagram channel config in the endpoints.yml.
> 
> With this I could successfully talk from facebook messenger and from instagram chat with the rasa chatbot.
> 
> Thoughts? @Horizon733 @alwx

Answer for Point 1:
The echo message is sent by Instagram, and it might be a bug, i don't know. For quick fix I added a condition so bot doesn't loop around with its own message -> response.
Answer for Point 2:
When you are using same facebook page for Instagram and facebook messenger. Then you can use `both` as keyword for using the channel connector for both services.

rgstephens commented:
> the singular facebook webhook handles the IG messages?

Yes

JEM-Mosig commented on rasa/nlu/extractors/entity_synonyms.py:
```diff
@@ -13,16 +12,17 @@
 from rasa.engine.storage.resource import Resource
 from rasa.engine.storage.storage import ModelStorage
 
+# This is a workaround around until we have all components migrated to `GraphComponent`.
+EntitySynonymMapper = EntitySynonymMapper
```
It looks like this is not up to date with main. Once you update, this should not touch any Research-owned code and you don't need a review from Research (unless it still does).

Horizon733 commented:
@JEM-Mosig I was waiting for a review since 3.0 release. But after that too there was no response. So, I didn't check here (no offense). I will resolve those conflicts and push it till next week

Horizon733 commented:
Hi @TyDunn @alwx Happy new year. I have merged and solved conflicts with main branch.  Please review and let me know if there is anything to be done or changed

Horizon733 commented:
Also, I would like to tell you guys, I am not able to run livedocs since it gives me following error. I did check googling it was showing there might be path issue with windows machine.
````
(env) D:\rasa>make livedocs             
cd docs/ && poetry run yarn start

  FileNotFoundError

  [WinError 2] The system cannot find the file specified

  at E:\Anaconda3\lib\subprocess.py:1207 in _execute_child
      1203│                                          int(not close_fds),
      1204│                                          creationflags,
      1205│                                          env,
      1206│                                          os.fspath(cwd) if cwd is not None else None,
    → 1207│                                          startupinfo)
      1208│             finally:
      1209│                 # Child is launched. Close the parent's copy of those pipe
      1210│                 # handles that only the child should have open.  You need
      1211│                 # to make sure that no handles to the write end of the
Makefile:233: recipe for target 'livedocs' failed
make: *** [livedocs] Error 1
```

losterloh commented:
@Horizon733 Are you still interested in getting this reviewed & merged? 🙂 If yes, please take care of the conflicts and then we're happy to take another look!

Horizon733 commented:
> @Horizon733 Are you still interested in getting this reviewed & merged? 🙂 If yes, please take care of the conflicts and then we're happy to take another look!

Hi I am interested in merging and having this feature, but the reviewing and approval takes so long so this conflicts occur if see all merges above i have solve conflicts many times. I will do solve conflicts if it will be reviewed fast and provide me review. Because as Rasa is this big code are changes everytime in week.

losterloh commented:
@Horizon733 Yes, we will be happy to do a fast review!

Horizon733 commented:
> @Horizon733 Yes, we will be happy to do a fast review!

@losterloh I am facing following issue while pulling from the main branch. Can you please help me out?
![image](https://user-images.githubusercontent.com/57827233/172411818-1e9888f5-6b5d-4475-b401-9c96700d3cf4.png)

Horizon733 commented:
> @Horizon733 Yes, we will be happy to do a fast review!

@losterloh Can you please review the code and let me know

sanchariGr commented:
Hey @Horizon733, your changes look good but would be great if you can add some tests along with it, as a reference we have the existing channel tests for [facebook](https://github.com/RasaHQ/rasa/blob/main/rasa/core/channels/facebook.py) or any of the other supported channels can be found [here](https://github.com/RasaHQ/rasa/tree/main/rasa/core/channels). Feel free to take a look and let us know how you feel about this and if you need some more infos.

Horizon733 commented:
> Hey @Horizon733, your changes look good but would be great if you can add some tests along with it, as a reference we have the existing channel tests for [facebook](https://github.com/RasaHQ/rasa/blob/main/rasa/core/channels/facebook.py) or any of the other supported channels can be found [here](https://github.com/RasaHQ/rasa/tree/main/rasa/core/channels). Feel free to take a look and let us know how you feel about this and if you need some more infos.

Hi @sanchariGr  Sorry for late reply, I am writing tests, I will definitely try to push it by this week's end.

Horizon733 commented:
Hi @TyDunn @sanchariGr I am facing below error because of pre-commit and black can you guys please help me out committing and pushing changes so we can move ahead?
```
(.venv) F:\rasa>git commit -m "tests for messenger added"
black....................................................................Failed
- hook id: black
- exit code: 1

Traceback (most recent call last):
  File "D:\miniconda3\envs\rasa\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "D:\miniconda3\envs\rasa\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\Scripts\black.EXE\__main__.py", line 7, in <module>
  File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\black\__init__.py", line 1129, in patched_main
    patch_click()
  File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\black\__init__.py", line 1115, in patch_click
    from click import _unicodefun  # type: ignore
ImportError: cannot import name '_unicodefun' from 'click' (C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\click\__init__.py)

doctoc...............................................(no files to check)Skipped
docstring-check..........................................................Failed
- hook id: docstring-check
- exit code: 2

git diff main...HEAD -- rasa | poetry run flake8 --select D --diff
rasa/core/channels/messenger.py:25:1: D102 Missing docstring in public method
rasa/core/channels/messenger.py:28:1: D107 Missing docstring in __init__
rasa/core/channels/messenger.py:175:1: D102 Missing docstring in public method
rasa/core/channels/messenger.py:178:1: D107 Missing docstring in __init__
rasa/core/channels/messenger.py:187:1: D202 No blank lines allowed after function docstring
rasa/core/channels/messenger.py:294:1: D202 No blank lines allowed after function docstring
rasa/core/channels/messenger.py:333:1: D102 Missing docstring in public method
rasa/core/channels/messenger.py:337:1: D102 Missing docstring in public method
rasa/core/channels/messenger.py:360:1: D417 Missing argument descriptions in the docstring
rasa/core/channels/messenger.py:378:1: D102 Missing docstring in public method
rasa/core/channels/messenger.py:452:1: D102 Missing docstring in public method
make: *** [Makefile:95: lint-docstrings] Error 1
```
The docstrings are similar to which we have in facebook.py channel code not any addon or any removal just typos have been changed then too it is cancelling commits

sanchariGr commented:
> Hi @TyDunn @sanchariGr I am facing below error because of pre-commit and black can you guys please help me out committing and pushing changes so we can move ahead?
> 
> ```
> (.venv) F:\rasa>git commit -m "tests for messenger added"
> black....................................................................Failed
> - hook id: black
> - exit code: 1
> 
> Traceback (most recent call last):
>   File "D:\miniconda3\envs\rasa\lib\runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "D:\miniconda3\envs\rasa\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\Scripts\black.EXE\__main__.py", line 7, in <module>
>   File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\black\__init__.py", line 1129, in patched_main
>     patch_click()
>   File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\black\__init__.py", line 1115, in patch_click
>     from click import _unicodefun  # type: ignore
> ImportError: cannot import name '_unicodefun' from 'click' (C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\click\__init__.py)
> 
> doctoc...............................................(no files to check)Skipped
> docstring-check..........................................................Failed
> - hook id: docstring-check
> - exit code: 2
> 
> git diff main...HEAD -- rasa | poetry run flake8 --select D --diff
> rasa/core/channels/messenger.py:25:1: D102 Missing docstring in public method
> rasa/core/channels/messenger.py:28:1: D107 Missing docstring in __init__
> rasa/core/channels/messenger.py:175:1: D102 Missing docstring in public method
> rasa/core/channels/messenger.py:178:1: D107 Missing docstring in __init__
> rasa/core/channels/messenger.py:187:1: D202 No blank lines allowed after function docstring
> rasa/core/channels/messenger.py:294:1: D202 No blank lines allowed after function docstring
> rasa/core/channels/messenger.py:333:1: D102 Missing docstring in public method
> rasa/core/channels/messenger.py:337:1: D102 Missing docstring in public method
> rasa/core/channels/messenger.py:360:1: D417 Missing argument descriptions in the docstring
> rasa/core/channels/messenger.py:378:1: D102 Missing docstring in public method
> rasa/core/channels/messenger.py:452:1: D102 Missing docstring in public method
> make: *** [Makefile:95: lint-docstrings] Error 1
> ```
> 
> The docstrings are similar to which we have in facebook.py channel code not any addon or any removal just typos have been changed then too it is cancelling commits

Hey @Horizon733 looks to me like both flake8 and black are flagging some linting issues. Can you try fixing the issues you get after running `poetry run flake8 rasa tests --extend-ignore D` and `poetry run black rasa tests` and then try to commit.

Horizon733 commented:
> > Hi @TyDunn @sanchariGr I am facing below error because of pre-commit and black can you guys please help me out committing and pushing changes so we can move ahead?
> > ```
> > (.venv) F:\rasa>git commit -m "tests for messenger added"
> > black....................................................................Failed
> > - hook id: black
> > - exit code: 1
> > 
> > Traceback (most recent call last):
> >   File "D:\miniconda3\envs\rasa\lib\runpy.py", line 193, in _run_module_as_main
> >     "__main__", mod_spec)
> >   File "D:\miniconda3\envs\rasa\lib\runpy.py", line 85, in _run_code
> >     exec(code, run_globals)
> >   File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\Scripts\black.EXE\__main__.py", line 7, in <module>
> >   File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\black\__init__.py", line 1129, in patched_main
> >     patch_click()
> >   File "C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\black\__init__.py", line 1115, in patch_click
> >     from click import _unicodefun  # type: ignore
> > ImportError: cannot import name '_unicodefun' from 'click' (C:\Users\disha\.cache\pre-commit\repowotxool4\py_env-python3\lib\site-packages\click\__init__.py)
> > 
> > doctoc...............................................(no files to check)Skipped
> > docstring-check..........................................................Failed
> > - hook id: docstring-check
> > - exit code: 2
> > 
> > git diff main...HEAD -- rasa | poetry run flake8 --select D --diff
> > rasa/core/channels/messenger.py:25:1: D102 Missing docstring in public method
> > rasa/core/channels/messenger.py:28:1: D107 Missing docstring in __init__
> > rasa/core/channels/messenger.py:175:1: D102 Missing docstring in public method
> > rasa/core/channels/messenger.py:178:1: D107 Missing docstring in __init__
> > rasa/core/channels/messenger.py:187:1: D202 No blank lines allowed after function docstring
> > rasa/core/channels/messenger.py:294:1: D202 No blank lines allowed after function docstring
> > rasa/core/channels/messenger.py:333:1: D102 Missing docstring in public method
> > rasa/core/channels/messenger.py:337:1: D102 Missing docstring in public method
> > rasa/core/channels/messenger.py:360:1: D417 Missing argument descriptions in the docstring
> > rasa/core/channels/messenger.py:378:1: D102 Missing docstring in public method
> > rasa/core/channels/messenger.py:452:1: D102 Missing docstring in public method
> > make: *** [Makefile:95: lint-docstrings] Error 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The docstrings are similar to which we have in facebook.py channel code not any addon or any removal just typos have been changed then too it is cancelling commits
> 
> Hey @Horizon733 looks to me like both flake8 and black are flagging some linting issues. Can you try fixing the issues you get after running `poetry run flake8 rasa tests --extend-ignore D` and `poetry run black rasa tests` and then try to commit.

Hi @sanchariGr,
It seems commit happens from linux or mac os without M1 chip and has difficulties with Windows. I hope you guys can have some tips on how to run tests on windows and able to commit via windows.
I have added some updates inside `Docs` please do check them and let me know if there needs to be any changes, we can have discussion like what heading, subheading and content we can have for docs.
Apart from that, tests has been added, please do check those.
Thanks

sanchariGr commented:
@Horizon733 you might be facing this issue because your local version of black doesnot match the one on the CI, can you please try running `make formatter` and `make lint`?

Horizon733 commented:
> @Horizon733 you might be facing this issue because your local version of black doesnot match the one on the CI, can you please try running `make formatter` and `make lint`?

let me try and push the code

Horizon733 commented:
![image](https://user-images.githubusercontent.com/57827233/180241905-2da1d193-591c-4607-94d7-77249cd16094.png)
Hi @sanchariGr, the lines that lint specifies were/are already like that, (missing docstrings and have blank lines) so is there any way to ignore those, or what can we do? I am pushing the reformatted code, please try and let me know how it is on your end

sanchariGr commented on rasa/core/channels/messenger.py:
```diff
@@ -23,15 +23,17 @@ class Messenger:
     @classmethod
     def name(cls) -> Text:
-        return "facebook"
+        return "messenger"
```
```suggestion
        """Return channel name."""
        return "messenger"
```

sanchariGr commented on rasa/core/channels/messenger.py:
```diff
@@ -170,11 +173,14 @@ class MessengerBot(OutputChannel):
     @classmethod
     def name(cls) -> Text:
-        return "facebook"
+        return "messenger"
```
```suggestion
        """Return channel name."""
        return "messenger"
```

sanchariGr commented on rasa/core/channels/messenger.py:
```diff
@@ -170,11 +173,14 @@ class MessengerBot(OutputChannel):
+    def __init__(
+        self, messenger_client: MessengerClient, messenger_service: Text
+    ) -> None:
 
```
```suggestion
      """Initialises input channel."""
```

sanchariGr reviewed (commented):
@Horizon733 Unfortunately, the linter will flag any issues in the module that’s being edited, even though that particular code might not have been touched. The only way forward is for you to add docstrings. I have added some to help get started.

sanchariGr commented:
Also @Horizon733 can you also run all the unit tests locally with `make test` there are a couple other tests that were dependant on facebook.py that are currently [failing](https://github.com/RasaHQ/rasa/runs/7520888454?check_suite_focus=true), you probably just need to change it to the updated name.
